pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,243 @@
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCallbackFn),
+		fs:       &utilfs.DefaultFs{},
+	}
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Creates the plugin directory, if it doesn't already exist.
+func (w *Watcher) createPluginDir() error {
+	glog.Infof("Ensuring Plugin directory at %s ", w.path)
+	err := w.fs.MkdirAll(w.path, 0755)
+	if err != nil {
+		return fmt.Errorf("error (re-)creating driver directory: %s", err)
+	}
+	return nil
+}
+
+// Walks through the plugin directory to discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) init() error {
+	if err := w.createPluginDir(); err != nil {
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
+	if err != nil {
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := w.invokeRegistrationCallbackAtHandler(infoResp, socketPath)
+	glog.V(2).Infof("call NotifyRegistrationStatus, resp: %v", resp)
+	if _, err := client.NotifyRegistrationStatus(ctx, resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", infoResp.Type, infoResp.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", infoResp.Type, infoResp.Name, socketPath)
+	return nil
+}
+
+func (w *Watcher) invokeRegistrationCallbackAtHandler(infoResp *registerapi.PluginInfo, socketPath string) *registerapi.RegistrationStatus {
+	var handlerCbkFn RegisterCallbackFn
+	var ok bool
+	handlerCbkFn, ok = w.handlers[infoResp.Type]
```

Comment:
should check the plugin type? which defined at apis/pluginregistration/v1alpha1/constants.go

Reply:
I think this is what the next line does unless I didn't understand your comment :) ?

Reply:
no, if you look at the example_plugin.go, it using plugin type = "example-plugin-type",
basically registerapi.PluginInfo.Type is not been enforced

Reply:
Do you mean strictly enforce the plugins in constants.go? In that case I'm not sure that's needed.

Plugin type is enforced by the fact that the type is or isn't in the array. Adding to the array is a matter of registering a plugin, in the case of the test, it passes because a handler added with that type.

Reply:
right, i mean the first part, otherwise i don't see the purpose of defining plugin type const at constants.go

Reply:
can you please explain more on security worry part? `AddHandler` can only be invoked in a static manner from within the kubelet code.

Reply:
sorry guys, security is not a concern after some thought, the only improvement i can think of is: add log even when handler been invoked successfully at invokeRegistrationCallbackAtHandler(), this is for audit purpose base on AAA model,  this logging improvement can be applied on other functions as well.